src/controller/CommissioneeDeviceProxy.cpp

```diff
@@ -0,0 +1,329 @@
+            {
+                mState = ConnectionState::NotConnected;
+                ReturnErrorOnFailure(LoadSecureSessionParameters());
+                didLoad = true;
+            }
+        }
+        else
+        {
+            mState = ConnectionState::NotConnected;
+            ReturnErrorOnFailure(LoadSecureSessionParameters());
+            didLoad = true;
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CommissioneeDeviceProxy::SendCommands(app::CommandSender * commandObj)
+{
+    bool loadedSecureSession = false;
+    ReturnErrorOnFailure(LoadSecureSessionParametersIfNeeded(loadedSecureSession));
+    VerifyOrReturnError(commandObj != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    return commandObj->SendCommandRequest(mSecureSession.Value());
+}
+
+void CommissioneeDeviceProxy::OnNewConnection(SessionHandle session)
+{
+    mState = ConnectionState::SecureConnected;
+    mSecureSession.SetValue(session);
+
+    // Reset the message counters here because this is the first time we get a handle to the secure session.
+    // Since CHIPDevices can be serialized/deserialized in the middle of what is conceptually a single PASE session
+    // we need to restore the session counters along with the session information.
+    Transport::SecureSession * secureSession = mSessionManager->GetSecureSession(mSecureSession.Value());
+    VerifyOrReturn(secureSession != nullptr);
+    MessageCounter & localCounter = secureSession->GetSessionMessageCounter().GetLocalMessageCounter();
+    if (localCounter.SetCounter(mLocalMessageCounter) != CHIP_NO_ERROR)
+    {
+        ChipLogError(Controller, "Unable to restore local counter to %" PRIu32, mLocalMessageCounter);
+    }
+    Transport::PeerMessageCounter & peerCounter = secureSession->GetSessionMessageCounter().GetPeerMessageCounter();
+    peerCounter.SetCounter(mPeerMessageCounter);
+}
+
+void CommissioneeDeviceProxy::OnConnectionExpired(SessionHandle session)
+{
+    VerifyOrReturn(mSecureSession.HasValue() && mSecureSession.Value() == session,
+                   ChipLogDetail(Controller, "Connection expired, but it doesn't match the current session"));
+    mState = ConnectionState::NotConnected;
+    mSecureSession.ClearValue();
+}
+
+CHIP_ERROR CommissioneeDeviceProxy::OnMessageReceived(Messaging::ExchangeContext * exchange, const PayloadHeader & payloadHeader,
+                                                      System::PacketBufferHandle && msgBuf)
+{
+    if (mState == ConnectionState::SecureConnected)
+    {
+        if (mStatusDelegate != nullptr)
+        {
+            mStatusDelegate->OnMessage(std::move(msgBuf));
+        }
+        else
+        {
+            HandleDataModelMessage(exchange, std::move(msgBuf));
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+void CommissioneeDeviceProxy::OnResponseTimeout(Messaging::ExchangeContext * ec) {}
+
+CHIP_ERROR CommissioneeDeviceProxy::CloseSession()
```

Comment:
Does anything need to be done to mPairing here?

Reply:
Adding  `Clear()` to it. It's generally a good practice to clear the memory of these sessions, as it carries cryptographic information.